cmd/bench/bench.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"code.cloudfoundry.org/bytefmt"
+	"encoding/hex"
+	"flag"
+	"fmt"
+	"github.com/olekukonko/tablewriter"
+	"github.com/spacemeshos/post/initialization"
+	"github.com/spacemeshos/post/proving"
+	"github.com/spacemeshos/post/shared"
+	"github.com/spacemeshos/post/validation"
+	"log"
+	"math"
+	"os"
+	"runtime"
+	"strconv"
+	"time"
+)
+
+type Config = shared.Config
+
+var (
+	id, _         = hex.DecodeString("deadbeef")
+	challenge, _  = hex.DecodeString("this is a challenge")
+	defaultConfig = shared.DefaultConfig()
+)
+
+func main() {
+	datadir := flag.String("datadir", defaultConfig.DataDir, "filesystem datadir path")
+	space := flag.Uint64("space", 1<<23, "space per unit, in bytes")
+	single := flag.Bool("single", false, "whether to execute a single test instead of the complete set")
+	flag.Parse()
+
+	log.Printf("bench config: datadir: %v, space: %v", *datadir, bytefmt.ByteSize(*space))
+
+	cases := genTestCases(*datadir, *space, *single)
+	data := make([][]string, 0)
+	for i, cfg := range cases {
+		log.Printf("test %v/%v starting...", i+1, len(cases))
+		tStart := time.Now()
+
+		init := initialization.NewInitializer(&cfg, shared.DisabledLogger{})
+		prover := proving.NewProver(&cfg, shared.DisabledLogger{})
+		validator := validation.NewValidator(&cfg)
+
+		t := time.Now()
+		proof, err := init.Initialize(id)
+		if err != nil {
+			panic(err)
+		}
+		eInit := time.Since(t)
+
+		t = time.Now()
+		err = validator.Validate(proof)
+		if err != nil {
+			panic(err)
+		}
+		eInitV := time.Since(t)
```

Comment:
Since the validation is very short (sub ms), consider running it more than once? (e.g. loop 1000 times and divide the total time by 1000)

Also consider not rounding it to the ms so it's possible to notice differences in validation time.

Same for the other validation bench.

Reply:
I think that if result is below 1ms (and due to the rounding, if it's below 0.5ms), the result is meaningless anyway. 

But in order to avoid seeing `0ms` in the report, i'll round it up to a Microsecond.

What do you say?

Reply:
I'm trying to see how validation scales: we'll validate PoST proofs for all miners once per epoch, so let's say 5k miners. If it takes 1ms then: 1ms * 5k = 5s.

I agree that 5 seconds once per epoch isn't terrible, so I'm ok with only changing the rounding.